src/spi/bus.rs

```diff
@@ -0,0 +1,369 @@
+use crate::clock::Clocks;
+use crate::time::Hertz;
+use bare_metal::Mutex;
+use core::cell::RefCell;
+use core::convert::Infallible;
+use e310x::{QSPI0, QSPI1, QSPI2};
+pub use embedded_hal::spi::{Mode, Phase, Polarity, MODE_0, MODE_1, MODE_2, MODE_3};
+use nb;
+
+use super::{Pins, PinsNoSS, SpiX};
+
+/// SPI bus abstraction
+pub struct Spi<SPI, PINS> {
+    spi: SPI,
+    pins: PINS,
+    shared: bool, // indicate if this bus is shared
+}
+
+/// Cell<Spi> locked behind a Mutex
+pub type SharedBus<SPI, PINS> = Mutex<RefCell<Spi<SPI, PINS>>>;
+
+impl<SPI: SpiX, PINS> Spi<SPI, PINS> {
+    /// Construct the [Spi] as fully exclusive bus for direct use with a single device
+    /// in full duplex master mode.
+    /// Defaults to using HOLD CS in FRAME mode if PINS configuration allows it.
+    /// Constructing the [Spi] like this makes it exclusive to one device.
+    /// Use the [shared](Spi::shared) constructor to share the [Spi] bus accross devices.
+    pub fn new(spi: SPI, pins: PINS, mode: Mode, freq: Hertz, clocks: Clocks) -> Self
+    where
+        PINS: Pins<SPI>,
+    {
+        let mut result = Self {
+            spi,
+            pins,
+            shared: false,
+        };
+
+        let cs_index = PINS::CS_INDEX;
+
+        // derive cd_mode based on PINS composition by type when bus is used exclusively
+        let cs_mode = if cs_index.is_some() {
+            e310x::qspi0::csmode::MODE_A::HOLD
+        } else {
+            e310x::qspi0::csmode::MODE_A::OFF
+        };
+
+        let div = clocks.tlclk().0 / (2 * freq.0) - 1;
+
+        // configure the bus right away for exclusive use
+        result.configure(mode, div, cs_mode, cs_index);
+
+        result
+    }
+
+    /// Construct the [Spi] as uninitialized bus for use with [SpiSharedDevice](super::SpiSharedDevice).
+    /// Can only be used with non-CS [Pins] as the CS pins are used by each
+    /// [SpiSharedDevice](super::SpiSharedDevice) individually.
+    /// Constructs the struct behind a `Mutex<RefCell>` for interrupt safety.
+    pub fn shared(spi: SPI, pins: PINS) -> Mutex<RefCell<Self>>
+    where
+        PINS: PinsNoSS<SPI>,
+    {
+        Mutex::new(RefCell::new(Self {
+            spi,
+            pins,
+            shared: true,
+        }))
+    }
+
+    /// Indicate if this bus is to be shared accross multiple devices or
+    /// used exclusively for one.
+    pub fn is_shared(&self) -> bool {
```

Comment:
This method is probably useless because shared SPI has a completely different type.

Reply:
it's a convenience method but I'll break the API in my next update so we get 2 separate device types at which point it'll be useless.

Reply:
removed